crates/wasmtime/src/component/values.rs

```diff
@@ -0,0 +1,389 @@
+            }
+            Val::Record(Record { values, .. }) | Val::Tuple(Tuple { values, .. }) => {
+                let mut offset = offset;
+                for value in values.deref() {
+                    value.store(mem, value.ty().next_field(&mut offset))?;
+                }
+            }
+            Val::Variant(Variant {
+                discriminant,
+                value,
+                ty,
+            }) => self.store_variant(*discriminant, value, ty.cases().len(), mem, offset)?,
+
+            Val::Enum(Enum { discriminant, ty }) => {
+                self.store_variant(*discriminant, &Val::Unit, ty.names().len(), mem, offset)?
+            }
+
+            Val::Union(Union {
+                discriminant,
+                value,
+                ty,
+            }) => self.store_variant(*discriminant, value, ty.types().len(), mem, offset)?,
+
+            Val::Option(Option {
+                discriminant,
+                value,
+                ..
+            })
+            | Val::Expected(Expected {
+                discriminant,
+                value,
+                ..
+            }) => self.store_variant(*discriminant, value, 2, mem, offset)?,
+
+            Val::Flags(Flags { count, value, .. }) => {
+                match FlagsSize::from_count(*count as usize) {
+                    FlagsSize::Size1 => u8::try_from(value[0]).unwrap().store(mem, offset)?,
+                    FlagsSize::Size2 => u16::try_from(value[0]).unwrap().store(mem, offset)?,
+                    FlagsSize::Size4Plus(_) => {
+                        let mut offset = offset;
+                        for value in value.deref() {
+                            value.store(mem, offset)?;
+                            offset += 4;
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn store_variant<T>(
+        &self,
+        discriminant: u32,
+        value: &Val,
+        case_count: usize,
+        mem: &mut MemoryMut<'_, T>,
+        offset: usize,
+    ) -> Result<()> {
+        let discriminant_size = DiscriminantSize::from_count(case_count).unwrap();
+        match discriminant_size {
+            DiscriminantSize::Size1 => u8::try_from(discriminant).unwrap().store(mem, offset)?,
+            DiscriminantSize::Size2 => u16::try_from(discriminant).unwrap().store(mem, offset)?,
+            DiscriminantSize::Size4 => (discriminant).store(mem, offset)?,
+        }
+
+        value.store(
+            mem,
+            offset
+                + func::align_to(
+                    discriminant_size.into(),
+                    self.ty().size_and_alignment().alignment,
+                ),
+        )
+    }
+}
+
+/// Lower a list with the specified element type and values.
+fn lower_list<T>(mem: &mut MemoryMut<'_, T>, items: &[Val]) -> Result<(usize, usize)> {
+    let element_type = match items {
+        [item, ..] => item.ty(),
+        [] => Type::Unit,
```

Comment:
I think this is actually subtly incorrect because it would get the alignment incorrect for the call to `realloc`. The type of the list element may need to get passed in here.